contents/blog/how-we-improved-feature-flags-resiliency.md

```diff
@@ -0,0 +1,134 @@
+
+Is that all we need to do? Not quite. As we found the painful way, just because we can defend against the database going down, doesn't mean our responses will be fast.
+
+### Different ways Postgres can bork
+
+When you can't connect to the database, it's a quick operation: you tried connecting, you failed, you raise an error, and any system depending on you can quickly make a decision.
+
+However, what if the database isn't down, but just painfully slow? For example:
+	- you're deadlocked in a transaction.
+	- you have several transactions running at the same time, slowing the overall system down.
+	- when a potentially infinite loop is hammering the database hundreds of time a second.
+  
+To us, a database that can't quickly respond is equivalent to a database that is down. To defend against this, we introduced statement timeouts of one second for flag evaluations. This helps us quickly catch if things are slow, and defends against timing out.
+
+### Different ways PgBouncer can bork
+
+PgBouncer manages connections to our Postgres database, and things can go wrong here too.
+
+The most popular way PgBouncer fails is when there are too many clients waiting for a connection. In this case, you're connected to PgBouncer, but waiting for a database connection to free up so you can make your query.
+
+The default wait for this setting is two minutes, which is way too long for us. PostHog times out at 10 seconds, so we want flags to have a much smaller timeout here. Currently, this is globally set at one second for all our systems.
+
+
+### Separating out deployment for feature flags
+
+One other way to defend against the database being too slow is to have a dedicated read-replica for flag evaluation. We've setup a separate deployment for feature flags (so it doesn't go down with the app), and have created a separate read-replica so it doesn't hit statement timeouts because something else is hammering the database.
+
+This has made flags a lot more reliable.
+
+## When things don't go as expected
+
+One frustrating issue we found towards the end is that when the database is slow, even setting a statement timeout during a session can take too long, which means by the time we figure out the database is slow, it's too late.
+
+Currently, we're thinking of introducing a separate PgBouncer pool for feature flags, which has these timeouts set on the connection itself.
+
+More long term, having this sync python service which can't handle app level timeouts well has been troublesome. We're considering rewriting this with an async paradigm that's much more efficient and allows for easily setting timeouts in the app, so we don't have to rely on the database level settings, and can once again share the connection pool with the rest of the app, rather than creating separate sub-pools.
```

Comment:
```suggestion
In the long term, having this sync python service that can't handle app level timeouts well has been troublesome. We're considering rewriting this with an async paradigm that's more efficient and allows for easily setting timeouts in the app, so we don't have to rely on the database level settings, and can once again share the connection pool with the rest of the app, rather than creating separate sub-pools.
```